### PR TITLE
Calculate extrusion width %s as a function of nozzle width, not layer height

### PR DESCRIPTION
**Pull request by iFallUpHill**
image

**Reviewable!**

### CODE REVIEW

**Comment by jeremytodd1** [comment]
Yes yes yes. This is a huge one.

I believe that this single feature will get quite a few people still using SuperSlicer to switch to Orca.

Thanks for working on this! I'm very excited that this may be close to release.

**Comment by SoftFever** [comment]
@iFallUpHill 
Cheers, thanks for taking on this task :)
I found a few places seems not handled properly yet in my initial review.
e.g. in gcode.cpp:1788
```
           float outer_wall_line_width = print.default_region_config().outer_wall_line_width.value;
            if (outer_wall_line_width == 0.0) {
                float default_line_width = print.default_object_config().line_width.value;
                outer_wall_line_width = default_line_width == 0.0 ? m_config.nozzle_diameter.get_at(initial_non_support_extruder_id) : default_line_width;
            }
            Flow outer_wall_flow = Flow(outer_wall_line_width, m_config.layer_height, m_config.nozzle_diameter.get_at(initial_non_support_extruder_id));
            float outer_wall_speed = print.default_region_config().outer_wall_speed.value;
```
and in src\libslic3r\MultiMaterialSegmentation.cpp:1548

```
    auto layer_color_stat = [&layers = std::as_const(layers)](const size_t layer_idx, const size_t color_idx) -> LayerColorStat {
        LayerColorStat out;
        const Layer &layer = *layers[layer_idx];
        for (const LayerRegion *region : layer.regions())
            if (const PrintRegionConfig &config = region->region().config();
                // color_idx == 0 means "don't know" extruder aka the underlying extruder.
                // As this region may split existing regions, we collect statistics over all regions for color_idx == 0.
                color_idx == 0 || config.wall_filament == int(color_idx)) {
                //BBS: the extrusion line width is outer wall rather than inner wall
                out.extrusion_width     = std::max<float>(out.extrusion_width, float(config.outer_wall_line_width));
                out.top_shell_layers    = std::max<int>(out.top_shell_layers, config.top_shell_layers);
                out.bottom_shell_layers = std::max<int>(out.bottom_shell_layers, config.bottom_shell_layers);
                out.small_region_threshold = config.gap_infill_speed.value > 0 ?
                                             // Gap fill enabled. Enable a single line of 1/2 extrusion width.
                                             0.5f * float(config.outer_wall_line_width) :
                                             // Gap fill disabled. Enable two lines slightly overlapping.
                                             float(config.outer_wall_line_width) + 0.7f * Flow::rounded_rectangle_extrusion_spacing(float(config.outer_wall_line_width), float(layer.height));
                out.small_region_threshold = scaled<float>(out.small_region_threshold * 0.5f);
                out.extrusion_spacing = Flow::rounded_rectangle_extrusion_spacing(float(config.outer_wall_line_width), float(layer.height));
                ++ out.num_regions;
            }
        assert(out.num_regions > 0);
        out.extrusion_width = scaled<float>(out.extrusion_width);
        out.extrusion_spacing = scaled<float>(out.extrusion_spacing);
        return out;
    };
```

There should be more places to handle percents, you can search for the exact name of all these changed parameters through the whole code base to find them.

**Comment by iFallUpHill** [comment]
@SoftFever Right, I think I got them except for this one:

Does this block actually do anything? I couldn't find `outer_wall_volumetric_speed` being used anywhere.
```
//BBS: calculate the volumetric speed of outer wall. Ignore pre-object setting and multi-filament, and just use the default setting
        {
            float filament_max_volumetric_speed = m_config.option<ConfigOptionFloats>("filament_max_volumetric_speed")->get_at(initial_non_support_extruder_id);
            float outer_wall_line_width = print.default_region_config().outer_wall_line_width.value;
            if (outer_wall_line_width == 0.0) {
                float default_line_width = print.default_object_config().line_width.value;
                outer_wall_line_width = default_line_width == 0.0 ? m_config.nozzle_diameter.get_at(initial_non_support_extruder_id) : default_line_width;
            }
            Flow outer_wall_flow = Flow(outer_wall_line_width, m_config.layer_height, m_config.nozzle_diameter.get_at(initial_non_support_extruder_id));
            float outer_wall_speed = print.default_region_config().outer_wall_speed.value;
            outer_wall_volumetric_speed = outer_wall_speed * outer_wall_flow.mm3_per_mm();
            if (outer_wall_volumetric_speed > filament_max_volumetric_speed)
                outer_wall_volumetric_speed = filament_max_volumetric_speed;
            m_placeholder_parser.set("outer_wall_volumetric_speed", new ConfigOptionFloat(outer_wall_volumetric_speed));
        }
```

**Comment by iFallUpHill** [comment]
Got supports working again, there are a couple of extrusion widths that I think I need to handle still (marked with `// TODO for extrusion width PR`, but otherwise could you take a look now? I think I got the rest of them.

@SoftFever

**Comment by SoftFever** [comment]
> @SoftFever Got supports working again, there are a couple of extrusion widths that I think I need to handle still (marked with `// TODO for extrusion width PR`, but otherwise could you take a look now? I think I got the rest of them.

Sure thing.
Give 1 or 2 days to go through it.

**Comment by SoftFever** [comment]
I was previously worried about multi-extruder scenarios, but since Orca doesn't support multi-extruder yet, I think we can worry about that later.

File: src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -1057,14 +1057,17 @@ void PrintConfigDef::init_fff_params()
...
-    def->sidetext = L("mm");
+    def->tooltip = L("Line width of outer wall. If expressed as a %, it will be computed over the nozzle diameter.");
+    def->sidetext = L("mm or %");
+    def->ratio_over = "nozzle_diameter";
     def->min = 0;
+    def->max = 1000;
+    def->max_literal = 10;
     def->mode = comAdvanced;
-    def->set_default_value(new ConfigOptionFloat(0));
+    def->set_default_value(new ConfigOptionFloatOrPercent(0.4, false));
```
Comment (SoftFever) [inline src/libslic3r/PrintConfig.cpp:1070]: Would be better  to set default to `ConfigOptionFloatOrPercent(0, false)`.
Same for other parameters.

**Comment by SoftFever** [comment]
It's a placeholder. User can access this variable in their custom g-codes

> @SoftFever ~Right, I think I got them except for this one:~
> 
> Does this block actually do anything? I couldn't find `outer_wall_volumetric_speed` being used anywhere.
> 
> **Gcode.cpp**
> 
> ```
> //BBS: calculate the volumetric speed of outer wall. Ignore pre-object setting and multi-filament, and just use the default setting
>         {
>             float filament_max_volumetric_speed = m_config.option<ConfigOptionFloats>("filament_max_volumetric_speed")->get_at(initial_non_support_extruder_id);
>             float outer_wall_line_width = print.default_region_config().outer_wall_line_width.value;
>             if (outer_wall_line_width == 0.0) {
>                 float default_line_width = print.default_object_config().line_width.value;
>                 outer_wall_line_width = default_line_width == 0.0 ? m_config.nozzle_diameter.get_at(initial_non_support_extruder_id) : default_line_width;
>             }
>             Flow outer_wall_flow = Flow(outer_wall_line_width, m_config.layer_height, m_config.nozzle_diameter.get_at(initial_non_support_extruder_id));
>             float outer_wall_speed = print.default_region_config().outer_wall_speed.value;
>             outer_wall_volumetric_speed = outer_wall_speed * outer_wall_flow.mm3_per_mm();
>             if (outer_wall_volumetric_speed > filament_max_volumetric_speed)
>                 outer_wall_volumetric_speed = filament_max_volumetric_speed;
>             m_placeholder_parser.set("outer_wall_volumetric_speed", new ConfigOptionFloat(outer_wall_volumetric_speed));
>         }
> ```
> 
> **Edit:** Also pretty sure the supports don't calculate correctly anymore :'(

File: src/libslic3r/MultiMaterialSegmentation.cpp [hunk 2]
```diff
@@ -1554,14 +1554,15 @@ static inline std::vector<std::vector<ExPolygons>> mmu_segmentation_top_and_bott
                 // As this region may split existing regions, we collect statistics over all regions for color_idx == 0.
                 color_idx == 0 || config.wall_filament == int(color_idx)) {
                 //BBS: the extrusion line width is outer wall rather than inner wall
-                out.extrusion_width     = std::max<float>(out.extrusion_width, float(config.outer_wall_line_width));
+                double outer_wall_line_width = config.get_abs_value("outer_wall_line_width");
```
Comment (SoftFever) [inline src/libslic3r/MultiMaterialSegmentation.cpp:1557]: This probably won't work.
Need to pass nozzle diameter

**Comment by iFallUpHill** [comment]
Supports work and I fixed the above two comments!

File: src/libslic3r/PrintConfig.cpp (hunk 1 above)
Comment (iFallUpHill) [reply src/libslic3r/PrintConfig.cpp]: changed all except for line_width

**Comment by Dustmuffins** [comment]
Great feature! Thanks!

**Comment by SoftFever** [comment]
@iFallUpHill 
Overall, looks good.
I made some minor fixes/changes to your branch,
Please review: https://github.com/SoftFever/OrcaSlicer/pull/1578/commits/1e9a4e9b0296855eb8026c6e6b777ce2196ee12b

File: src/libslic3r/Config.cpp [hunk 3]
```diff
@@ -653,22 +653,52 @@ double ConfigBase::get_abs_value(const t_config_option_key &opt_key) const
...
-            throw NoDefinitionException(opt_key);
-        const ConfigOptionDef *opt_def = def->get(opt_key);
-        assert(opt_def != nullptr);
-        // Compute absolute value over the absolute value of the base option.
-        //FIXME there are some ratio_over chains, which end with empty ratio_with.
-        // For example, XXX_extrusion_width parameters are not handled by get_abs_value correctly.
-        return opt_def->ratio_over.empty() ? 0. :
-            static_cast<const ConfigOptionFloatOrPercent*>(raw_opt)->get_abs_value(this->get_abs_value(opt_def->ratio_over));
+        auto cofop = static_cast<const ConfigOptionFloatOrPercent*>(raw_opt);
+            if (cofop->value == 0 && boost::ends_with(opt_key, "_line_width")) {
```
[the remark this answers is not in the document]
Comment (iFallUpHill) [reply src/libslic3r/Config.cpp:673]: Ah, this was copied when porting over the checks from here so I didn't think much of it:
https://github.com/supermerill/SuperSlicer/blob/a0f34b1637af67f8493d6aa118a56fc762a43b8c/src/libslic3r/Config.cpp#L751

My understanding was that this was how to fallback to `line_width` for 0 valued `XXX_line_width` parameters 😅

**Comment by iFallUpHill** [comment]
Thanks for the cleanup, definitely looks a lot cleaner now -- only comment left is the one you mentioned about the `_line_width` check. Feel free to revise to unblock the merge! :D

File: src/libslic3r/Config.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (SoftFever) [reply src/libslic3r/Config.cpp:673]: Make sense :)

**Comment by SoftFever** [comment]
Look good.
Merged.
Thank for the great work.

**Pull request merged**